lib/hanami/repository.rb

```diff
@@ -118,769 +106,322 @@ module Hanami
-      #   require 'hanami/model'
-      #
-      #   class ProjectRepository
-      #     include Hanami::Repository
-      #
-      #     def self.cool
-      #       query do
-      #         where(language: 'ruby')
-      #       end
-      #     end
-      #
-      #     def self.not_cool
-      #       exclude cool
-      #     end
-      #   end
-      def exclude(query)
-        query.negate!
-        query
-      end
+    # Initialize a new instance
+    #
+    # @return [Hanami::Repository] the new instance
+    #
+    # @since x.x.x
+    def initialize
+      super(self.class.container)
+    end
 
-      # This is a method to check entity persited or not
-      #
-      # @param entity
-      # @return a boolean value
-      # @since 0.3.1
-      def _persisted?(entity)
-        !!entity.id
-      end
+    # Find by primary key
+    #
+    # @return [Hanami::Entity,NilClass] the entity, if found
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   repository = UserRepository.new
+    #   user       = repository.create(name: 'Luca')
+    #
+    #   user       = repository.find(user.id)
+    def find(id)
+      root.by_primary_key(id).as(:entity).one
+    end
 
-      # Update timestamps
-      #
-      # @param entity [Object, Hanami::Entity] the entity
-      #
-      # @api private
-      # @since 0.3.1
-      def _touch(entity)
-        now = Time.now.utc
+    # Return all the records for the relation
+    #
+    # @return [Array<Hanami::Entity>] all the entities
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   UserRepository.new.all
+    def all
+      root.as(:entity)
+    end
 
-        if _has_timestamp?(entity, :created_at)
-          entity.created_at ||= now
-        end
+    # Returns the first record for the relation
+    #
+    # @return [Hanami::Entity,NilClass] first entity, if any
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   UserRepository.new.first
+    def first
+      root.as(:entity).first
+    end
 
-        if _has_timestamp?(entity, :updated_at)
-          entity.updated_at = now
-        end
-      end
+    # Returns the last record for the relation
+    #
+    # @return [Hanami::Entity,NilClass] last entity, if any
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   UserRepository.new.last
+    def last
+      root.order(Model::Sql.desc(root.primary_key)).as(:entity).first
+    end
 
-      # Check if the given entity has the given timestamp
-      #
-      # @param entity [Object, Hanami::Entity] the entity
-      # @param timestamp [Symbol] the timestamp name
-      #
-      # @return [TrueClass,FalseClass]
-      #
-      # @api private
-      # @since 0.3.1
-      def _has_timestamp?(entity, timestamp)
-        entity.respond_to?(timestamp) &&
-          entity.respond_to?("#{ timestamp }=")
-      end
+    # Deletes all the records from the relation
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   UserRepository.new.clear
+    def clear
+      root.delete
```

Comment:
Should `#as` and `#to_a` be called here to materialize the results and prevent chaining?